snappy/security.go

```
@@ -121,106 +329,616 @@ func getSecurityProfile(m *packageYaml, appName, baseDir string) (string, error)
+		if err != nil {
+			logger.Noticef("Failed to generate seccomp policy for %s: %v", name, err)
+			return nil, err
 		}
-		if s.PolicyVendor != "" {
-			policyVendor = s.PolicyVendor
+	}
+	res.scFn = filepath.Join(dirs.SnapSeccompDir, res.id.AppID)
+	res.aaFn = filepath.Join(dirs.SnapAppArmorDir, res.id.AppID)
+
+	return res, nil
+}
+
+func (sd *SecurityDefinitions) generatePolicyForServiceBinary(m *packageYaml, name string, baseDir string) error {
+	p, err := sd.generatePolicyForServiceBinaryResult(m, name, baseDir)
+	if err != nil {
+		return err
+	}
+
+	os.MkdirAll(filepath.Dir(p.scFn), 0755)
+	err = ioutil.WriteFile(p.scFn, []byte(p.scPolicy), 0644)
+	if err != nil {
+		logger.Noticef("Failed to write seccomp policy for %s: %v", name, err)
+		return err
+	}
+
+	os.MkdirAll(filepath.Dir(p.aaFn), 0755)
+	err = ioutil.WriteFile(p.aaFn, []byte(p.aaPolicy), 0644)
+	if err != nil {
+		logger.Noticef("Failed to write AppArmor policy for %s: %v", name, err)
+		return err
+	}
+	out, err := loadAppArmorPolicy(p.aaFn)
+	if err != nil {
+		logger.Noticef("Failed to load AppArmor policy for %s: %v\n:%s", name, err, out)
+		return err
+	}
+
+	return nil
+}
+
+// FIXME: move into something more generic - SnapPart.HasConfig?
+func hasConfig(baseDir string) bool {
+	return helpers.FileExists(filepath.Join(baseDir, "meta", "hooks", "config"))
+}
+
+func generatePolicy(m *packageYaml, baseDir string) error {
+	var foundError error
+
+	// generate default security config for snappy-config
+	if hasConfig(baseDir) {
+		if err := snappyConfig.generatePolicyForServiceBinary(m, "snappy-config", baseDir); err != nil {
+			foundError = err
+			logger.Noticef("Failed to obtain APP_ID for %s: %v", "snappy-config", err)
 		}
-		if s.PolicyVersion != 0 {
-			policyVersion = s.PolicyVersion
+	}
+
+	for _, service := range m.ServiceYamls {
+		err := service.generatePolicyForServiceBinary(m, service.Name, baseDir)
+		if err != nil {
+			foundError = err
```

Same as the other note: this will hide all errors but the last one.